design/metric-math.md

```diff
@@ -0,0 +1,202 @@
+  * Risk of 'us' introducing bugs is minimized
+
+### Cons
+* More verbose - the developer has to write a lot of boiler-plate such as allocating ids, returnData, etc.
+* The developer must learn how to express valid metric expressions according to the somewhat complicated API and documentation. The `returnData` nuances were particulary difficult to tease out from the docs, for example.
+* Not type-safe - there is no static checking of the expression to ensure it is valid. Failures are caught at deployment time which is very slow, unless we build our own parser ... ?
+
+## Option 2 - Slight Enhancement
+
+Straying from the original CFN format a little, we can specialize the API in a way that better reflects the usual case:
+1. Fetch and assign metrics to IDs
+2. Use an expression to compute and return a `TS` result
+
+```typescript
+const errors = new cloudwatch.Metric({
+  metricName: 'Errors',
+  namespace: 'Test'
+});
+const requests = new cloudwatch.Metric({
+  metricName: 'Requests',
+  namespace: 'Test'
+});
+
+const alarm = new cloudwatch.Alarm(this, 'woof', {
+  // etc.
+  metrics: {
+    ids: {
+      errors,
+      requests
+    },
+    expression: 'errors / requests * 100'
+  }
+});
+```
+
+### Pros
+* Compact
+  * It takes less code to express the math
+  * There is no need to manage `returnData` which is not as useful for alarms as it is for the `GetMetricData` API.
+
+### Cons
+* Different to the original CFN and SDK API.
+* It's opinionated - are there metrics that can not be expressed in this way?
+* Also not statically checked - errors in the expression will be found at deployment time.
+
+## Option 3 - Type-safe DSL
+
+Model the data types, operators and functions as classes and methods in code.
+
+```typescript
+const errors = new cloudwatch.Metric({
+  metricName: 'Errors',
+  namespace: 'Test'
+});
+const requests = new cloudwatch.Metric({
+  metricName: 'Requests',
+  namespace: 'Test'
+});
+
+const alarm = new cloudwatch.Alarm(this, 'woof', {
+  // etc.
+  metrics: errors.divide(requests).multiply(100)
+});
+```
+
+### Pros
+* Type-safe - the compiler and implementation checks the validity of expressions. For example, it ensures the final result is a `TS` and the arguments to functions or operators are used correctly.
+* IDE discoverability, auto-complete, in-line documentation.
+* Programmatic variation - like how we build constructs, the expression tree can be assembled incrementally with standard programming techniques (`if-else`, `functions`, `classes`, etc.)
+
+### Cons
+* Highly opinionated and different to the original API
+* We must maintain backwards compatibility with any future features released by the CloudWatch team
+* Risk of us introducing bugs is higher - especially for large complicated expressions
+* Expressing paranthesis to control mathematical precedence rules may be ugly or unintuitive
+* Could be problematic for JSII, although it does pass its checks :)
+
```

Comment:
Might be also interesting to explore something between option 2 and option 3:

```ts
const errors = new cloudwatch.Metric(...);
const requests = new cloudwatch.Metric(...);

const alarm = new cloudwatch.Alarm(this, 'woof', {
  expression: `${errors} / ${requests} * 100`
});
```

Pros:
- Utilizes the CloudWatch syntax, which makes it easier for people to look up help, docs, examples, etc.
- Concise

Cons:
- Not type-safe
- Magic all around

Reply:
Interesting idea. Are you thinking that `toString()` of `Metric` would encode its information as a string so that `Alarm` can parse the expression and construct the JSON payload? Similar to how Tokens work now?

Reply:
Exactly. We can probably even use the existing token stringification mechanism to achieve this. It's designed to be quite general purpose (albeit a bit complex).